### ISSUE TITLE
Split verification is broken?

### ISSUE THREAD

**Issue by sitsofe** [axboe/fio#801]
With master of fio (currently fc220349e45144360917db48010b503a9874930d) I see a `verify: bad header numberio 1, wanted 0 at file /tmp/fio.tmp offset 4096, length 4096 (requested block: offset=4096, length=4096)` error when trying to do a split verification job using `verify_only`:

```
$ ./fio --minimal --name=go --verify=crc32c --filename=/tmp/fio.tmp --size=8k --rw=write 
3;fio-3.15-2-gfc2203;go;0;0;8;8000;2000;1;0;0;0.000000;0.000000;2;10;6.297000;5.794033;1.000000%=2;5.000000%=2;10.000000%=2;20.000000%=2;30.000000%=2;40.000000%=2;50.000000%=2;60.000000%=10;70.000000%=10;80.000000%=10;90.000000%=10;95.000000%=10;99.000000%=10;99.500000%=10;99.900000%=10;99.950000%=10;99.990000%=10;0%=0;0%=0;0%=0;2;10;6.437000;5.878886;0;0;0.000000%;0.000000;0.000000;8;8000;2000;1;0;0;0.000000;0.000000;7;39;23.204500;22.795001;1.000000%=7;5.000000%=7;10.000000%=7;20.000000%=7;30.000000%=7;40.000000%=7;50.000000%=7;60.000000%=39;70.000000%=39;80.000000%=39;90.000000%=39;95.000000%=39;99.000000%=39;99.500000%=39;99.900000%=39;99.950000%=39;99.990000%=39;0%=0;0%=0;0%=0;10;65;38.481500;38.874610;0;0;0.000000%;0.000000;0.000000;0.000000%;0.000000%;0;0;15;100.0%;0.0%;0.0%;0.0%;0.0%;0.0%;0.0%;0.00%;25.00%;25.00%;25.00%;25.00%;0.00%;0.00%;0.00%;0.00%;0.00%;0.00%;0.00%;0.00%;0.00%;0.00%;0.00%;0.00%;0.00%;0.00%;0.00%;0.00%;0.00%;dm-0;0;0;0;0;0;0;0;0.00%;slaves;0;7;0;0;0;0;0;0.00%;sda;0;7;0;0;0;0;0;0.00%
$ ./fio --minimal --name=go --verify=crc32c --filename=/tmp/fio.tmp --size=8k --rw=write --verify_only=1 --verify_state_load=1
verify: bad header numberio 1, wanted 0 at file /tmp/fio.tmp offset 4096, length 4096 (requested block: offset=4096, length=4096)
fio: pid=6492, err=84/file:io_u.c:2029, func=io_u_sync_complete, error=Invalid or incomplete multibyte or wide character
3;fio-3.15-2-gfc2203;go;0;84;8;8000;2000;1;0;0;0.000000;0.000000;5;380;193.040000;264.595115;1.000000%=5;5.000000%=5;10.000000%=5;20.000000%=5;30.000000%=5;40.000000%=5;50.000000%=5;60.000000%=378;70.000000%=378;80.000000%=378;90.000000%=378;95.000000%=378;99.000000%=378;99.500000%=378;99.900000%=378;99.950000%=378;99.990000%=378;0%=0;0%=0;0%=0;6;380;193.572000;264.885029;0;0;0.000000%;0.000000;0.000000;8;0;0;0;0;0;0.000000;0.000000;129624;129630;129627.880000;4.285067;1.000000%=129499;5.000000%=129499;10.000000%=129499;20.000000%=129499;30.000000%=129499;40.000000%=129499;50.000000%=129499;60.000000%=129499;70.000000%=129499;80.000000%=129499;90.000000%=129499;95.000000%=129499;99.000000%=129499;99.500000%=129499;99.900000%=129499;99.950000%=129499;99.990000%=129499;0%=0;0%=0;0%=0;0;24;12.377000;16.536399;0;0;0.000000%;0.000000;0.000000;0.000000%;0.000000%;2;0;30;100.0%;0.0%;0.0%;0.0%;0.0%;0.0%;0.0%;0.00%;0.00%;25.00%;0.00%;0.00%;0.00%;0.00%;25.00%;0.00%;0.00%;0.00%;0.00%;0.00%;0.00%;0.00%;0.00%;50.00%;0.00%;0.00%;0.00%;0.00%;0.00%;dm-0;0;0;0;0;0;0;0;0.00%;slaves;1;6;0;0;0;0;0;0.00%;sda;1;6;0;0;0;0;0;0.00%
```

I think jobs of this type worked in the distant past (e.g. fio 2.18?) but I think they fail since fio 3.1..

**Comment by axboe** [comment]
Can you try and bisect it?

**Comment by sitsofe** [comment]
Sure, I'll post more when I've done so.

**Comment by sitsofe** [comment]
I was wrong about fio 3.1 not working for verification - it works as well as fio-2.18 but the difference was it output a
```
layout: No I/O performed by psync, perhaps try --debug=io option for details?
```
message too (but the verification itself actually happens).

I started another bisect and narrowed the problem down to 5c5c33c1b0df088187a6d4fac5459df5e029763a (fio 3.6 era):
```
$ git bisect bad 
5c5c33c1b0df088187a6d4fac5459df5e029763a is the first bad commit
commit 5c5c33c1b0df088187a6d4fac5459df5e029763a
Author: Bart Van Assche <bart.vanassche@wdc.com>
Date:   Mon Mar 26 09:00:28 2018 -0700

    Only populate the write buffer if necessary
    
    This patch moves the populate_verify_io_u() call from inside
    get_io_u() into all its callers except do_dry_run() and thereby skips
    write buffer population for dry runs. This patch does not change the
    behavior of fio but is necessary because the ZBC patch will insert
    code in fill_io_u() after the get_io_u() call that may modify the
    write offset.
    
    Signed-off-by: Bart Van Assche <bart.vanassche@wdc.com>

:100644 100644 f2d7cc368f0f63534218fc60e13d7ef895b0e3c5 b28c3dbc4fd65fdf71b3ad4581a9b27e9aac7511 M	backend.c
:100644 100644 d4a65089a502ece5f28d58ef38ad84ae85def358 5fbb2386237de6ba8aa106917e9f222ae564e014 M	io_u.c
$ git bisect log 
git bisect start
# good: [f2cd91604af170e972438c461a40230e266a57d9] debug: fix inverted logic in fio_did_warn()
git bisect good f2cd91604af170e972438c461a40230e266a57d9
# bad: [fc220349e45144360917db48010b503a9874930d] Merge branch 'dev' of https://github.com/smartxworks/fio
git bisect bad fc220349e45144360917db48010b503a9874930d
# bad: [49e1ade14949b737dbd111b19a4057faf9aa77bb] axmap: return early of an overlap results in 0 settable bits
git bisect bad 49e1ade14949b737dbd111b19a4057faf9aa77bb
# bad: [50d2a37543908e79c41b1d5a4c2cb2d65c5af21b] make fio scripts python3-ready (part 2)
git bisect bad 50d2a37543908e79c41b1d5a4c2cb2d65c5af21b
# good: [87c6f22bf24da4679849ccf778451b8432c2b368] server: use scalloc() for sk_out allocation
git bisect good 87c6f22bf24da4679849ccf778451b8432c2b368
# bad: [f4b49c5523f14f39ecae250c64f8f93a9e3c1855] iolog: fix issue with replay rate
git bisect bad f4b49c5523f14f39ecae250c64f8f93a9e3c1855
# bad: [465964a6c8ff3ccac62b92e9af57377075e04579] Merge branch 'master' of https://github.com/bvanassche/fio
git bisect bad 465964a6c8ff3ccac62b92e9af57377075e04579
# good: [196ce5d5ff76af30cf93fd24240f12d8c4d24381] Merge branch 'fixbug_glfs' of https://github.com/simon-rock/fio
git bisect good 196ce5d5ff76af30cf93fd24240f12d8c4d24381
# bad: [9109883a1c20c2cbadd2d2cb8ca80d03835eaa66] option parsing: Mark arguments that are not modified as 'const'
git bisect bad 9109883a1c20c2cbadd2d2cb8ca80d03835eaa66
# good: [553f965256ab58b0e0958b0e367d8e63698f7a26] Rename TD_F_VER_NONE into TD_F_DO_VERIFY
git bisect good 553f965256ab58b0e0958b0e367d8e63698f7a26
# bad: [69349699d5a6e244fd9e74665ff2962cbca3d329] parse.h: Remove a superfluous cast
git bisect bad 69349699d5a6e244fd9e74665ff2962cbca3d329
# bad: [5c5c33c1b0df088187a6d4fac5459df5e029763a] Only populate the write buffer if necessary
git bisect bad 5c5c33c1b0df088187a6d4fac5459df5e029763a
# first bad commit: [5c5c33c1b0df088187a6d4fac5459df5e029763a] Only populate the write buffer if necessary
```

**Comment by sitsofe** [comment]
I suspect this is fixed by 8859c0675aec03467ed65dfc79ecc874d026f165 ...

**Comment by sitsofe** [comment]
The test case above now works without error so I'm closing this issue.
